subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec.groovy

```diff
@@ -163,12 +163,54 @@ public class MyTest {
         |}""".stripMargin()
 
         and:
+        succeeds ":$task"
+
+        then:
+        executedAndNotSkipped ":$task"
+
+        where:
+        suiteName   | suiteDeclaration              | task        | type
+        'test'      | 'test'                        | 'test'      | 'includeCategories'
+        'test'      | 'test'                        | 'test'      | 'excludeCategories'
+        'integTest' | 'integTest(JvmTestSuite)'     | 'integTest' | 'includeCategories'
+        'integTest' | 'integTest(JvmTestSuite)'     | 'integTest' | 'excludeCategories'
+    }
+
+    @Issue('https://github.com/gradle/gradle/issues/4924')
+    def "skips test on re-run when #type is NOT changed"() {
+        given:
+        resources.maybeCopy("JUnitCategoriesIntegrationSpec/reExecutesWhenPropertyIsChanged")
+        buildFile << """
+        |testing {
+        |   suites {
+        |       test {
+        |           useJUnit()
+        |           targets {
+        |               all {
+        |                   testTask.configure {
+        |                       options {
+        |                           ${type} 'org.gradle.CategoryA'
+        |                       }
+        |                   }
+        |               }
+        |           }
+        |       }
+        |   }
+        |}""".stripMargin()
+
+        when:
         succeeds ':test'
 
         then:
         executedAndNotSkipped ':test'
 
+        when:
+        succeeds ':test'
+
+        then:
+        skipped ':test'
+
         where:
-        type << ['includeCategories', 'excludeCategories']
+        type << ['includeCategories']
```

Comment:
Think we can remove this `where` all-together now

Reply:
I think I'll restore the `excludeCategories` here actually - I just removed it to make debugging quicker - so that this test mirrors the other pre-existing test, which checked both.